### DISCUSSION TITLE
Compiler error of missing traits when using .into_boxed(), BoxedQuery, QueryDsl, and .count()

### DISCUSSION THREAD

**Discussion by HubertMajewski** [diesel-rs/diesel#4374] [Q&A]
### Versions

- **Rust:** 1.81
- **Diesel:** 2.2.5
- **Database:** PostgreSQL
- **Operating System** Arch Linux

### Feature Flags

- **diesel:** postgres_backend

## Problem Description
While creating a wrapper for serializing data contents & their totals, .count() statements (and their extending .select() statements) run a compile-time error for missing Table, AsQuery, Select<CountStar> for QueryDsl through the where clause.

### What are you trying to accomplish?
I was looking into creating a wrapper to implement paginate responses. Thus using .into_boxed() for applying dynamic filters & sortings.

### What is the expected output?

No Compiler error of missing traits.

Some common documentation I found about using the corresponding types say to use QueryDsl. Using said type does not cause errors on .into_boxed() or proceeding with .filter() but does with .select() or .count().

For example, SelectDsl:
```
The into_boxed method

This trait should not be relied on directly by most apps. Its behavior is provided by QueryDsl. However, you may need a where clause on this trait to call into_boxed from generic code.
```

### What is the actual output?
Compiler error stating missing traits:

```
the trait bound `<T as MyWrapperImpl>::BoxedType: Table` is not satisfied
required for `<T as MyWrapperImpl>::BoxedType` to implement `SelectDsl<CountStar>`rustc
main.rs(57, 25): consider further restricting the associated type: ` where <T as MyWrapperImpl>::BoxedType: Table`
the trait bound `<T as MyWrapperImpl>::BoxedType: Query` is not satisfied
required for `<T as MyWrapperImpl>::BoxedType` to implement `AsQuery`
required for `<T as MyWrapperImpl>::BoxedType` to implement `SelectDsl<CountStar>`
main.rs(57, 25): consider further restricting the associated type: ` where <T as MyWrapperImpl>::BoxedType: Query`
```

### Steps to reproduce

main.rs
```
use diesel::prelude::*;

diesel::table! {
    my_db {
        id -> BigInt,
        first_name -> Text
    }
}

#[derive(Debug, Identifiable, Insertable)]
#[diesel(table_name = my_db)]
struct MyStruct {
    #[diesel(skip_insertion)]
    id: i64,
    first_name: String,
}

trait MyWrapperImpl
where
    Self::BoxedType: QueryDsl,
{
    type BoxedType;

    fn get_boxed() -> Self::BoxedType;

    fn apply_filters(query: Self::BoxedType) -> Self::BoxedType;
}

struct MyWrapper<T> {
    total: usize,
    data: Vec<T>
}

use diesel::associations::HasTable;
impl MyWrapperImpl for MyStruct
where
    Self: HasTable,
{
    type BoxedType = my_db::BoxedQuery<'static, diesel::pg::Pg>;

    fn get_boxed() -> Self::BoxedType {
        Self::table().into_boxed()
    }
    
    fn apply_filters(mut query: Self::BoxedType) -> Self::BoxedType {
        query = query.filter(my_db::first_name.eq("testName"));
        
        return query;
    }

}

impl<T> MyWrapper<T> 
where
    T: MyWrapperImpl
{
    pub fn new() -> Self {
        let mut test = T::get_boxed();

        test = T::apply_filters(test);

        let count = test.count(); // Errors Here

        Self {
            total: 0,
            data: Vec::new()
        }
    }
}

fn main() {
    let query = my_db::table.into_boxed();

    let total = query.count(); // Compiles Here
}

```

## Checklist

- [X] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues) and the [discussion forum](https://github.com/diesel-rs/diesel/discussions) for similar possible closed issues.
 
- [X] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)
- [X] This issue can be reproduced without requiring a third party crate

**Comment by weiznich** [comment]
Thanks for filling this. I've converted this to a discussion thread entry as this is not a bug report but a support question.

Now the problem here is that your generic code is missing required trait bounds, so rustc cannot resolve the correct methods there. I personally do generally not provide complete solutions in such cases as that often lead to situations where people just copy code without understanding it, which in turn makes it harder to refactor that code later on (which in turn brings the same people here again asking me to update their code).

Now the problem here is that this line 

>```rust
>        let mut test = T::get_boxed();
> ```

construct something of the type `T::BoxedType`, where the compiler only knows that `T::BoxedType: QueryDsl` is true. The documentation of [`QueryDsl::count`](https://docs.diesel.rs/2.2.x/diesel/prelude/trait.QueryDsl.html#method.count) list additional bounds that need to be fulfilled (`Self: SelectDsl<CountStar>`), which the compiler does not know about, therefore it generates an error message.

As for the specific error message: That one is not great, but there is not much we can do here until `#[diagnostic::do_not_recommend]` is stable. Given that attribute this message would rather read like: `<T as MyWrapperImpl>::BoxedType: SelectDsl<CountStar> is not satisfied`.
